File: app/src/main/java/de/davis/passwordmanager/dashboard/SecureElementDiffCallback.java
package de.davis.passwordmanager.dashboard;

import androidx.recyclerview.widget.DiffUtil;

import java.util.List;

import de.davis.passwordmanager.security.element.SecureElement;

public class SecureElementDiffCallback extends DiffUtil.Callback {

    private final List<Item> oldItems;
    private final List<Item> newItems;

    public SecureElementDiffCallback(List<Item> oldItems, List<Item> newItems) {
        this.oldItems = oldItems;
        this.newItems = newItems;
    }

    @Override
    public int getOldListSize() {
        return oldItems.size();
    }

    @Override
    public int getNewListSize() {
        return newItems.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        return oldItems.get(oldItemPosition).getId() == newItems.get(newItemPosition).getId();
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        Item oldItem = oldItems.get(oldItemPosition);
        Item newItem = newItems.get(newItemPosition);

        if(oldItem instanceof SecureElement && newItem instanceof SecureElement){
            return ((SecureElement) oldItem).getUniqueString().equals(((SecureElement) newItem).getUniqueString());
        }

        return false;
    }
}

File: app/src/main/java/de/davis/passwordmanager/dashboard/SecureElementDiffCallback.kt
package de.davis.passwordmanager.dashboard

import androidx.recyclerview.widget.DiffUtil
import de.davis.passwordmanager.database.dtos.SecureElement

class SecureElementDiffCallback(
    private val oldItems: List<Item>,
    private val newItems: List<Item>
) : DiffUtil.Callback() {
    override fun getOldListSize(): Int {
        return oldItems.size
    }

    override fun getNewListSize(): Int {
        return newItems.size
    }

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return oldItems[oldItemPosition].id == newItems[newItemPosition].id
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        val oldItem = oldItems[oldItemPosition]
        val newItem = newItems[newItemPosition]
        return if (oldItem is SecureElement && newItem is SecureElement) {
            oldItem == newItem
        } else false
    }
}